### ISSUE TITLE
Example page object for user list

### ISSUE THREAD

**Issue by christianacca** [andresdominguez/protractor-codelab#2]
Hi @andresdominguez,

I'm creating a page object for the user list view (#/exercise3). 

For this scenario **do you think the following makes a good page object**:

**Protractor test:**

``` typescript
    it('should find the email address for John Rambo', function() {
        var ramboEmailDisplay = screen
            .userRowsByName("John Rambo")
            .map((row) => Exercise2Screen.asUserRow(row).email.getText());

        expect(ramboEmailDisplay).toEqual(["rambo@gmail.com"])
    });
```

**Page object:**

``` typescript
export default class Exercise2Screen {
    userRows = element.all(by.repeater("row in ctrl.userList"));
    open() {
        browser.get("#/exercise2")
    };

    userRowsByName(name: string) {
        return this.userRows.filter((row) => {
            return Exercise2Screen.asUserRow(row).name.getText().then((n) => n === name)
        })
    };
    static asUserRow = (row: protractor.ElementFinder) => {
        return new UserRow(row);
    };
}

class UserRow {
    constructor(private row: protractor.ElementFinder) { }
    name = this.row.element(by.binding("row.name"));
    email = this.row.element(by.binding("row.email"));
    phone = this.row.element(by.binding("row.phone"));
}
```

**Comment by andresdominguez** [comment]
Hello @christianacca 

I usually write my page object by looking at the test. I start with an empty test and then I start asking myself: what kind of page object **represents the user actions** and **reads in plain english**.

For example:

Step 1. How does the flow looks like from the user's point of view?

``` js
it('should find the email address for John Rambo', function() {
  // Navigate to list page
  // Find row by name
  // Get the email of the row
  // Ensure the email is "rambo@gmail.com"
});
```

Step 2: Start writing functions that would achieve step1 and read in plain english

``` js
it('should find the email address for John Rambo', function() {
  // Navigate to list page
  listPage.get();

  // Find row by name, get email
  var email = listPage.findRowByName('John Rambo').getEmail();

  // Ensure the email is "rambo@gmail.com"
  expect(email).toBe('rambo@gmail.com');
});
```

Notice that there is no logic in the test. There are no maps, forEach, filters.

Now that you have a test that looks good to you , start writing your page object. What do you need to implement in order to make your test work?

Maybe you can do something like this:

``` js
class ListPage {
  constructor() {
  }

  get() {
    browser.get('...');
  },

  private getRow(name) {
    return $$('.user-row').filter(function(row) {
      return row.element(by.binding('row.name')).getText().then(function(rName) {
        return name === rName;
      });
    });
  },

  findRowByName(name) {
    var row = getRow(name);
    return {
      getEmail: function() {
        return row.then(rows => {
          return rows[0].element(by.binding('row.email')).getText();
        });
      }
    }
  }
}
```

The page object implementation is not as important as your test. Your test is the most important part. It should be short, easy to read, and it should serve as documentation.

**Comment by christianacca** [comment]
@andresdominguez, that's great advice.

One objective that's important for me to ensure that a page object does not define too narrow an API, and yet have the actual tests read as clearly as possible.

I also want a consistent API for the page objects so that the DSL becomes a common vocabulary.

Here's what I've come up with:

**Protractor tests**

``` typescript
    it('should find the email address for John Rambo', () => {
        var email = screen.userRows.find(UserRow.where.nameEq("John Rambo")).email;

        expect(email.getText()).toEqual("rambo@gmail.com")
    });
```

**Page object**

``` typescript
export class Exercise2Screen {
    userRows = new UserRowList(element.all(by.repeater("row in ctrl.userList")))
    open() {
        browser.get("#/exercise2")
    };
}

class UserRowList extends RowList<UserRow, UserRowList> {

    create(rows: protractor.ElementArrayFinder) {
        return new UserRowList(rows);
    }

    wrap(row: protractor.ElementFinder) {
        return new UserRow(row);
    };
}


export class UserRow {
    constructor(private row: protractor.ElementFinder) { }
    name = this.row.element(by.binding("row.name"));
    email = this.row.element(by.binding("row.email"));
    phone = this.row.element(by.binding("row.phone"));
    static where = {
        nameEq(name: string): (row: UserRow) => webdriver.promise.Promise<boolean> {
            return (r: UserRow) => r.name.getText().then((n) => n === name);
        }
    }
}

abstract class RowList<RowT, ListT> {
    constructor(public rows: protractor.ElementArrayFinder) { }

    count() {
        return this.rows.count();
    }

    abstract create(rows: protractor.ElementArrayFinder) : RowList<RowT, ListT>

    filter<T>(predicate: (row: RowT) => webdriver.promise.Promise<boolean>): RowList<RowT, ListT> {
        var matches = this.rows.filter((row) => {
            var userRow = this.wrap(row);
            return predicate(userRow);
        });
        return this.create(matches);
    }

    find<T>(predicate: (row: RowT) => webdriver.promise.Promise<boolean>): RowT {
        return this.filter(predicate).first();
    }

    first() {
        return this.wrap(this.rows.first());
    }
    get(index: number) {
        return this.wrap(this.rows.get(index));
    };

    abstract wrap(row: protractor.ElementFinder) : RowT;
}

```
